### PR TITLE
Initial Implementation of Stream and NSOutputStream

### PR DESCRIPTION
**Pull request by saiHemak**

### CODE REVIEW

**Comment by saiHemak** [comment]
This and https://github.com/apple/swift-corelibs-foundation/pull/442  have  common changes in 
 TestFoundation/TestNSStream.swift

TestFoundation/main.swift

Foundation/NSStream.swift

**Comment by saiHemak** [comment]
Adding TestNStream.swift to project.pbxproj has been taken care in the PR #442.

File: Foundation/NSStream.swift [hunk 1]
```diff
@@ -66,11 +69,11 @@ public class Stream: NSObject {
     }
     
     public func open() {
-        NSUnimplemented()
+        streamOpen()
```
Comment (phausler) [inline Foundation/NSStream.swift]: Please refer to the comment on #442 per requiring subclassers to implement these

File: Foundation/NSStream.swift [hunk 2]
```diff
@@ -138,35 +155,54 @@ public class InputStream: Stream {
 // Subclassers are required to implement these methods.
 // Currently this is left as named NSOutputStream due to conflicts with the standard library's text streaming target protocol named OutputStream (which ideally should be renamed)
 public class NSOutputStream : Stream {
+    
+    private  var _stream: CFWriteStream!
+    
     // writes the bytes from the specified buffer to the stream up to len bytes. Returns the number of bytes actually written.
     public func write(_ buffer: UnsafePointer<UInt8>, maxLength len: Int) -> Int {
-        NSUnimplemented()
+        if let outputStream = _stream {
```
Comment (phausler) [inline Foundation/NSStream.swift]: You should be able to bang on this because it only should be nil during the initialization

**Comment by saiHemak** [comment]
@phausler : Thanks  for your review .I have addressed your comments ..Please review

**Comment by phausler** [comment]
@swift-ci Please test

**Comment by saiHemak** [comment]
I have resolved the conflicts ..

File: Foundation/NSStream.swift [hunk 3]
```diff
@@ -155,35 +155,50 @@ public class InputStream: Stream {
...
-
+    
     public init?(url: URL, append shouldAppend: Bool) {
-        NSUnimplemented()
+        _stream = CFWriteStreamCreateWithFile(kCFAllocatorSystemDefault, url._cfObject)
     }
     
     public convenience init?(toFileAtPath path: String, append shouldAppend: Bool) {
-        NSUnimplemented()
+        self.init(url: URL(fileURLWithPath: path), append: true)
```
Comment (phausler) [inline Foundation/NSStream.swift]: Why I this always appending?

File: Foundation/NSStream.swift [hunk 4]
```diff
@@ -155,35 +155,50 @@ public class InputStream: Stream {
...
 
     public init(toBuffer buffer: UnsafeMutablePointer<UInt8>, capacity: Int) {
-        NSUnimplemented()
+        _stream = CFWriteStreamCreateWithBuffer(kCFAllocatorSystemDefault, buffer, capacity)
     }
-
+    
     public init?(url: URL, append shouldAppend: Bool) {
-        NSUnimplemented()
+        _stream = CFWriteStreamCreateWithFile(kCFAllocatorSystemDefault, url._cfObject)
```
Comment (phausler) [inline Foundation/NSStream.swift]: If the append is set this should set the stream property to append.

```obj-c
CFWriteStreamSetProperty((CFWriteStreamRef)object, kCFStreamPropertyAppendToFile, kCFBooleanTrue);
```

File: Foundation/NSStream.swift [hunk 5]
```diff
@@ -155,35 +155,50 @@ public class InputStream: Stream {
...
     }
     
     // returns YES if the stream can be written to or if it is impossible to tell without actually doing the write.
     public var hasSpaceAvailable: Bool {
-        NSUnimplemented()
+        return CFWriteStreamCanAcceptBytes(_stream)
     }
     
     public init(toMemory: ()) {
         NSUnimplemented()
```
Comment (phausler) [inline Foundation/NSStream.swift]: This one is basically this:

```objc
CFWriteStreamCreateWithAllocatedBuffers(NULL, kCFAllocatorDefault)
```

File: Foundation/NSStream.swift [hunk 6]
```diff
@@ -155,35 +155,50 @@ public class InputStream: Stream {
...
+        _stream = CFWriteStreamCreateWithFile(kCFAllocatorSystemDefault, url._cfObject)
     }
     
     public convenience init?(toFileAtPath path: String, append shouldAppend: Bool) {
-        NSUnimplemented()
+        self.init(url: URL(fileURLWithPath: path), append: true)
     }
     
     public class func outputStreamToMemory() -> Self {
         NSUnimplemented()
```
Comment (phausler) [inline Foundation/NSStream.swift]: This one is basically:

```objc
[[[self allocWithZone:NULL] initToMemory] autorelease]
```

File: Foundation/NSStream.swift (hunk 3 above)
Comment (saiHemak) [inline Foundation/NSStream.swift]: Modified the code to pass 'shouldAppend'  to self.init(url ..) which handles setting the stream property

File: Foundation/NSStream.swift (hunk 4 above)
Comment (saiHemak) [inline Foundation/NSStream.swift]: Added CFWriteStreamSetProperty() to set the writeStream property  after stream initialization .Also overridden the setPropertForKey() for 'writeStream'  which can be used to set stream property  after stream initialization.

File: Foundation/NSStream.swift (hunk 5 above)
Comment (saiHemak) [inline Foundation/NSStream.swift]: Implemented init(toMemory: ())

File: Foundation/NSStream.swift (hunk 6 above)
Comment (saiHemak) [inline Foundation/NSStream.swift]: Implemented outputStreamToMemory() however  the compiler mandated init(toMemory:()) to be marked as a ```required```  initializer

**Comment by saiHemak** [comment]
@phausler :Thanks for the review and comments.I have responded to all of them.I have also contributed tests that covers most of the contributed code.I have taken this up as a prerequisite for implementing  ```NSJSONSerialization.writeJSONObject(toStream)``` .I plan to add more tests to this class at a later point in time .

**Comment by phausler** [comment]
@swift-ci please test

**Pull request merged**